lib/system/services.toit

```diff
@@ -50,25 +49,52 @@ abstract class ServiceClient:
   abstract open -> ServiceClient?
 
   open_ uuid/string major/int minor/int -> ServiceClient?
-      --pid/int?=null
       --timeout/Duration?=_default_timeout_:
-    if _id_: throw "Already opened"
-    if pid:
-      process_send_ pid SYSTEM_RPC_NOTIFY_ [SERVICES_MANAGER_NOTIFY_ADD_PROCESS, Process.current.id]
+    return open_ uuid major minor --timeout=timeout --filter=: true
+
+  open_ uuid/string major/int minor/int [--filter] -> ServiceClient?
+      --timeout/Duration?=_default_timeout_:
+    discovered/List? := null
+    if timeout:
+      catch --unwind=(: it != DEADLINE_EXCEEDED_ERROR):
+        with_timeout timeout: discovered = _client_.discover uuid true
     else:
-      if timeout:
-        catch --unwind=(: it != DEADLINE_EXCEEDED_ERROR):
-          with_timeout timeout: pid = _client_.discover uuid true
-      else:
-        pid = _client_.discover uuid false
-      if not pid: return null
+      discovered = _client_.discover uuid false
+    if not discovered: return null
+
+    candidate_index := null
+    candidate_priority := null
+    for i := 0; i < discovered.size; i += 4:
+      tags := discovered[i + 3]
+      if filter.call tags:
```

Comment:
I'll probably change this so the filter receives the name, version, and tags independently -- possibly wrapped in a nice object.

Reply:
Done.